### PR TITLE
[luci] Propagate origin in optimization pass

### PR DESCRIPTION
**Pull request by mhs4670go**
This commit enables each optimization passes to propagate origins.

The number of files to review is kind of big but it simply uses origin setting API. Nevertheless, I think you will have trouble checking if `add_origin` api is used right way because you should know the entire workflow of its optimization. So, it would be better to take a look at the passes you implemented only.

FYI, only operators has its origin. It means the nodes like constant or bias can be ignored in this review.

Related: #6080 
Draft: #6217 
ONE-DCO-1.0-Signed-off-by: seongwoo <mhs4670go@naver.com>

### CODE REVIEW

**Review by llFreetimell — DISMISSED** [review]
LGTM

**Review by jinevening — DISMISSED** [review]
LGTM

File: compiler/luci/pass/src/FuseAddWithTConvPass.cpp [hunk 1]
```diff
@@ -105,6 +107,9 @@ bool fuse_add_with_tconv(luci::CircleTransposeConv *tconv)
     replace(add).with(tconv);
   }
 
+  // set origin
+  luci::add_origin(tconv, luci::get_origin(add));
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseAddWithTConvPass.cpp:111]: Q) I thought new nodes are set with origin. `tconv` is existing node. is there a rule for setting origin where I can understand the logic?

**Comment by seanshpark** [comment]
Some use `get_origin` and some uses `composite_origin`, both having one argument.
Is the any descript to understand the rule for this?

File: compiler/luci/pass/src/FuseAddWithTConvPass.cpp (hunk 1 above)
Comment (mhs4670go) [reply compiler/luci/pass/src/FuseAddWithTConvPass.cpp:111]: @seanshpark Basically, origin propagation is for preserving original nodes' information. If node A and node B are fused into new node C, node C should be inherited node A's and node B's origin. Likewise, if node A is fused to node B(no new node, just node B absorbed node A), then node B should get node A's origin. "preserving original nodes" would be helpful.

**Comment by mhs4670go** [comment]
@seanshpark Oh.. maybe `FuseBCQ.cpp` has kind of different logic as @llFreetimell implemented this file.
```
luci::add_origin(bcq_gather, luci::composite_origin({luci::get_origin(gather)}));
```
could be changed to
```
luci::add_origin(bcq_gather, luci::get_origin(gather));
```
I'll fix this part.
---
```
void add_origin(luci::CircleNode *circle_node, const std::shared_ptr<CircleNodeOrigin> origin)
```
And, FYI, `get_origin` is just a getter and `composite_origin` is a constructor-like method. `add_origin` needs `origin` as a second argument that will be added to given `circle_node`'s(first argument) origin. For second argument, we need to create a origin first. `single_origin` and `composite_origin` exist exactly for that purpose.

File: compiler/luci/pass/src/FuseAddWithTConvPass.cpp (hunk 1 above)
Comment (seanshpark) [reply compiler/luci/pass/src/FuseAddWithTConvPass.cpp:111]: Thanks for the explanation. It seems reasonable.
And is there any document for readers or guide for writers to follow?

**Review by llFreetimell — APPROVED** [review]
LGTM, thanks!

File: compiler/luci/pass/src/FuseAddWithTConvPass.cpp (hunk 1 above)
Comment (seanshpark) [reply compiler/luci/pass/src/FuseAddWithTConvPass.cpp:111]: > Well, no. Actually, it's a bit hard to find proper place to document.``)

Ah.. how about in `luci/Profile/CircleNodeOrigin.h` file? as this is another task, you can do this later or post an task issue not to forget :)

File: compiler/luci/pass/src/FuseBatchNormWithConvPass.cpp [hunk 2]
```diff
@@ -200,6 +201,8 @@ bool fused_batch_norm_with_conv(luci::CircleAdd *add)
   fused_conv->dilation()->h(conv->dilation()->h());
   fused_conv->dilation()->w(conv->dilation()->w());
   fused_conv->name(name + "/Conv2D");
+  luci::add_origin(fused_conv, luci::composite_origin({luci::get_origin(add), luci::get_origin(mul),
+                                                       luci::get_origin(conv)}));
```
Comment (seanshpark) [inline compiler/luci/pass/src/FuseBatchNormWithConvPass.cpp:205]: Q) just a question, is this equivalent? for understanding...
```cxx
luci::add_origin(fused_conv, luci::get_origin(add));
luci::add_origin(fused_conv, luci::get_origin(mul));
luci::add_origin(fused_conv, luci::get_origin(conv));
```

**Review by seanshpark — APPROVED** [review]
LGTM

File: compiler/luci/pass/src/FuseBatchNormWithConvPass.cpp (hunk 2 above)
Comment (mhs4670go) [reply compiler/luci/pass/src/FuseBatchNormWithConvPass.cpp:205]: @seanshpark Yes:)

**Review by jinevening — APPROVED** [review]
LGTM

**Review by struss — APPROVED** [review]
LGTM

**Pull request merged**